File: composeApp/src/androidMain/kotlin/com/github/jvsena42/loopky/ui/backup/BackupRingScreen.kt
package com.github.jvsena42.loopky.ui.backup

import android.content.Intent
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.net.toUri
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.github.jvsena42.loopky.R
import com.github.jvsena42.loopky.presentation.backup.BackupEffect
import com.github.jvsena42.loopky.presentation.backup.BackupRingUiState
import com.github.jvsena42.loopky.presentation.backup.BackupRingViewModel
import com.github.jvsena42.loopky.ui.components.LoopkyOutlinedButton
import com.github.jvsena42.loopky.ui.components.LoopkyPrimaryButton
import com.github.jvsena42.loopky.ui.signup.SignupScaffold
import com.github.jvsena42.loopky.ui.theme.LoopkyTheme
import com.github.jvsena42.loopky.ui.util.SecureScreen
import kotlinx.coroutines.flow.collectLatest
import org.koin.compose.viewmodel.koinViewModel

/**
 * Export the key into Pubky Ring over `pubkyring://`.
 *
 * Ring **imports** the key; it does not take custody. Loopky keeps its copy and its session, and
 * the phrase the user wrote down is still valid — so nothing on this screen may suggest the key
 * moved or that another copy stopped existing.
 */
@Composable
fun BackupRingRoute(
    onBack: () -> Unit,
    onDone: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: BackupRingViewModel = koinViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val currentOnDone by rememberUpdatedState(onDone)
    val context = LocalContext.current

    SecureScreen()

    LaunchedEffect(viewModel) {
        viewModel.effects.collectLatest { effect ->
            when (effect) {
                is BackupEffect.OpenDeeplink -> {
                    // Straight to startActivity. The URL carries the recovery phrase, so it is
                    // never logged and never staged in a clipboard — recents snapshots and IME
                    // clipboard history both read anything that passes through one.
                    val intent = Intent(Intent.ACTION_VIEW, effect.url.toUri())
                    if (intent.resolveActivity(context.packageManager) != null) {
                        context.startActivity(intent)
                    } else {
                        viewModel.onRingUnavailable()
                    }
                }
                BackupEffect.Done -> currentOnDone()
                else -> Unit
            }
        }
    }

    BackupRingScreen(
        state = state,
        onExport = viewModel::onExportClick,
        onConfirm = viewModel::onExportConfirmed,
        onBack = onBack,
        modifier = modifier,
    )
}

@Composable
private fun BackupRingScreen(
    state: BackupRingUiState,
    onExport: () -> Unit,
    onConfirm: () -> Unit,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LoopkyTheme.colors
    SignupScaffold(
        title = stringResource(R.string.backup_ring_title),
        subtitle = stringResource(R.string.backup_ring_subtitle),
        onBack = onBack,
        modifier = modifier,
    ) {
        if (!state.ringInstalled) {
            Text(
                text = stringResource(R.string.backup_ring_missing),
                modifier = Modifier.testTag("backup_ring_missing"),
                color = colors.foregroundSecondary,
                fontSize = 13.sp,
            )
            Spacer(Modifier.height(16.dp))
        }

        LoopkyPrimaryButton(
            label = stringResource(R.string.backup_ring_open),
            onClick = onExport,
            enabled = state.ringInstalled,
            modifier = Modifier.testTag("backup_ring_open"),
        )
        Spacer(Modifier.height(12.dp))
        // We cannot see the other side of the deeplink, so this is the user's word for it — which
        // is why it is a separate, deliberate tap rather than something inferred from the launch.
        LoopkyOutlinedButton(
            label = stringResource(R.string.backup_ring_confirm),
            onClick = onConfirm,
            modifier = Modifier.testTag("backup_ring_confirm"),
        )
    }
}

@Preview
@Composable
private fun BackupRingPreview() {
    LoopkyTheme {
        BackupRingScreen(
            state = BackupRingUiState(ringInstalled = true),
            onExport = {}, onConfirm = {}, onBack = {},
        )
    }
}

File: iosApp/iosApp/Views/BackupRingScreen.swift
import SwiftUI
import Shared

/// Export the key into Pubky Ring.
///
/// **Loopky keeps its copy.** Ring imports the key; it does not take custody of it, and a phrase
/// already written down stays valid.
struct BackupRingScreen: View {
    var onBack: () -> Void
    var onDone: () -> Void

    @Environment(\.openURL) private var openURL
    @Environment(\.scenePhase) private var scenePhase

    @State private var viewModel: BackupRingViewModel?
    @State private var uiState: BackupRingUiState?
    @State private var stateSink: FlowEffectSink?
    @State private var effectSink: FlowEffectSink?
    /// Only offered once the export link has actually been opened — the confirm is the user's
    /// word about what happened on the other side, and there is nothing to confirm before that.
    @State private var didExport = false

    var body: some View {
        SignupScaffold(
            title: "backup_ring_title",
            subtitle: NSLocalizedString("backup_ring_subtitle", comment: ""),
            errorTitle: uiState?.failed ?? false
                ? NSLocalizedString("restore_error_unreadable_title", comment: "")
                : nil,
            errorMessage: nil,
            onBack: onBack
        ) {
            VStack(alignment: .leading, spacing: 12) {
                if uiState?.ringInstalled ?? false {
                    SignupPrimaryButton(title: "backup_ring_open") {
                        viewModel?.onExportClick()
                    }
                    .accessibilityIdentifier("backup_ring_open")

                    // A separate tap, deliberately: we cannot see whether Ring accepted the key,
                    // so marking it backed up on the open would record a backup that may not
                    // exist.
                    if didExport {
                        SignupPrimaryButton(title: "backup_ring_confirm") {
                            viewModel?.onExportConfirmed()
                        }
                        .accessibilityIdentifier("backup_ring_confirm")
                    }
                } else {
                    Text("backup_ring_missing")
                        .font(.system(size: 14))
                        .foregroundStyle(LoopkyColor.foregroundSecondary)
                        .fixedSize(horizontal: false, vertical: true)
                    SignupPrimaryButton(title: "backup_ring_install") {
                        viewModel?.onInstallRingClick()
                    }
                    .accessibilityIdentifier("backup_ring_install")
                }
            }
        }
        .onAppear { attach() }
        .onDisappear { detach() }
        .onChange(of: scenePhase) { _, phase in
            // Re-probe on return from the App Store, so installing Ring does not leave the screen
            // insisting it is missing.
            if phase == .active { viewModel?.onScreenResumed() }
        }
    }

    private func attach() {
        guard viewModel == nil else { return }
        let vm = IosDependencies.shared.backupRingViewModel()
        viewModel = vm
        stateSink = FlowEffectSink(vm.state) { uiState = $0 as? BackupRingUiState }
        effectSink = FlowEffectSink(vm.effects) { effect in
            switch effect {
            case let deeplink as BackupEffectOpenDeeplink:
                // Straight out to the OS. **Never logged and never staged in a clipboard** — this
                // URL carries the recovery phrase.
                // Confirm is offered only if the link actually opened; a Ring that is not there
                // has exported nothing to confirm.
                if open(deeplink.url) {
                    didExport = true
                } else {
                    viewModel?.onRingUnavailable()
                }
            case let install as BackupEffectOpenInstallPage:
                _ = open(install.url)
            case is BackupEffectDone:
                onDone()
            default:
                break
            }
        }
    }

    @discardableResult
    private func open(_ raw: String) -> Bool {
        guard let url = URL(string: raw), UIApplication.shared.canOpenURL(url) else { return false }
        openURL(url)
        return true
    }

    private func detach() {
        viewModel?.release()
        viewModel = nil
        stateSink = nil
        effectSink = nil
    }
}
